digdag-standards/src/main/java/io/digdag/standards/command/SimpleCommandExecutor.java

```diff
@@ -1,24 +1,82 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Throwables;
 import com.google.inject.Inject;
-import com.google.common.base.Optional;
+import io.digdag.client.config.ConfigException;
 import io.digdag.spi.CommandExecutor;
-import io.digdag.spi.TaskRequest;
+import io.digdag.spi.CommandExecutorContext;
+import io.digdag.spi.CommandExecutorRequest;
+import io.digdag.spi.CommandLogger;
+import io.digdag.spi.CommandStatus;
+import io.digdag.spi.PrivilegedVariables;
+import java.io.IOException;
+import java.util.Map;
+import java.util.regex.Pattern;
 
 public class SimpleCommandExecutor
-    implements CommandExecutor
+        implements CommandExecutor
 {
+    private final static Pattern VALID_ENV_KEY = Pattern.compile("[a-zA-Z_][a-zA-Z_0-9]*");
+
+    private final CommandLogger clog;
+
     @Inject
-    public SimpleCommandExecutor()
-    { }
+    public SimpleCommandExecutor(final CommandLogger clog)
+    {
+        this.clog = clog;
+    }
 
-    public Process start(Path projectPath, TaskRequest request, ProcessBuilder pb)
-        throws IOException
+    @Override
+    public CommandStatus run(final CommandExecutorContext context, final CommandExecutorRequest request)
+            throws IOException
     {
+        final ProcessBuilder pb = new ProcessBuilder(request.getCommand());
+        pb.directory(context.getLocalProjectPath().toFile());
+        pb.redirectErrorStream(true);
+        pb.environment().putAll(request.getEnvironments());
+
         // TODO set TZ environment variable
-        return pb.start();
+        final Process p = pb.start();
+
+        // copy stdout to System.out and logger
+        clog.copyStdout(p, System.out);
+
+        // Need waiting and blocking. Because the process is running on a single instance.
+        // The command task could not be taken by other digdag-servers on other instances.
+        try {
+            p.waitFor();
+        }
+        catch (InterruptedException e) {
+            throw Throwables.propagate(e);
+        }
+
+        return SimpleCommandStatus.of(request.getIoDirectory().toString(), p);
+    }
+
+    /**
+     * This method is never called. The status of the task that is executed by the executor cannot be
+     * polled by non-blocking.
+     */
+    @Override
+    public CommandStatus poll(final CommandExecutorContext context, final ObjectNode previousStatusJson)
+            throws IOException
+    {
+        throw new UnsupportedOperationException("This method is never called.");
+    }
+
+    public static void collectEnvironmentVariables(final Map<String, String> env, final PrivilegedVariables variables)
+    {
+        for (String name : variables.getKeys()) {
+            if (!VALID_ENV_KEY.matcher(name).matches()) {
+                throw new ConfigException("Invalid _env key name: " + name);
+            }
+            env.put(name, variables.get(name));
+        }
+    }
+
+    public static boolean isValidEnvKey(String key)
```

Comment:
I suspect that VALID_ENV_KEY pattern and PrivilegedVariables are concerns of operator factories. Having this method here may enforce operator factories to follow this? But I haven't validated whether the pattern is most flexible on Windows and UNIX. I think it makes more sense to move this to digdag-plugin-utils, or keep it owned by one of operator factories.

Reply:
ditto https://github.com/treasure-data/digdag/pull/835/commits/e53857a94d5c7d1f89c9c36bb8d85a3a57438677